Skip bulk mapping step

allyhawkins opened the pull request:
Closes #86. This PR adds the flexibility to skip the salmon mapping step in the bulk workflow, similar to skipping the `salmon alevin` mapping in the alevin-fry workflow. In doing this, I implemented the suggestion in #86 to use a parameter `--repeat_mapping` and set the default value to `false`, meaning that mapping will be skipped by default if the necessary salmon output files are present and will be repeated if using the `--repeat_mapping` flag at the command line. 

I followed a very similar structure to the steps we take to skip mapping in the fry workflow, adding the salmon directory to the metadata and then using `branch` to split the channel based on having or not having the expected output file from salmon. The branch that doesn't have the salmon files done yet are then run through the `fastp` and `salmon` processes, while the other branch skips those processes. The output from the `salmon` process is then mixed with the channel that skipped mapping before creating the final output files. 

The main difference here is that instead of simply looking for the salmon output directory to be created (like we do in the fry workflow), I am looking for the creation of the `quant.sf` file which is explicitly needed in the next steps. Do reviewers agree with this decision or prefer to just look for the output directory? 

Additionally, here I am only looking for completion of the salmon step and then if that step is completed skipping both `fastp` and `salmon`. I thought it was fairly safe to assume that if the `quant.sf` files existed, then both processes had been previously completed and we did not need a separate check for if trimming had completed but not salmon since trimming is not very time consuming. 

I tested that this works both with skipping and not skipping mapping and all looked good. I also tested the fry workflow since I had changed the use of the parameter there and that also worked as expected.

jashapiro commented on modules/bulk-salmon.nf:
```diff
@@ -88,23 +88,46 @@ workflow bulk_quant_rna {
+          // add salmon directory and salmon file location to meta 
+          .map{it.salmon_publish_dir = "${params.outdir}/internal/salmon";
+               it.salmon_results_dir = "${it.salmon_publish_dir}/${it.library_id}";
+               it.salmon_quant_file = "${it.salmon_results_dir}/quant.sf";
```
I don't think we should need to check for the specific file, unless we are going to check for _all_ the files we will need, which would include the `.json` files as well. In most cases I would expect the directory should be sufficient. 

I did notice though that may be "leaking" the trimmed fastq files here, which should be deleted. I think this is because at line 13 the trimmed files are going into the same directory that we are using for the output. We should update that to maybe something like the below to avoid this.
```
trimmed_reads = "${meta.library_id}_trimmed"
```

jashapiro commented on modules/bulk-salmon.nf:
```diff
@@ -88,23 +88,46 @@ workflow bulk_quant_rna {
 
     emit: 
-        merge_bulk_quants.out
+        merge_bulk_quants.out.bulk_counts
```
Is there a reason you went back to separating these outputs? Is it required? If so, I suppose we probably should rename them as you had before, even if it looks redundant!

allyhawkins replied:
Oh yes, I forgot to mention that in the PR. I got an error that it couldn't emit a multi-channel output so had to change this. I can go back to naming them.

allyhawkins commented:
I went ahead and updated the changes so they only look for creation of the the salmon directory and added the `_trimmed` ending to the folder for the trimmed fastq files like you suggested. Only the salmon results should be included in that directory now.

jashapiro reviewed (approved):
LGTM! I did some testing and things seemed to work as expected. I did run into a failure for memory limits, so I thought we might bump up the config for the salmon step to match alevin. I don't think we should need to worry about larger disks, as I hope nobody is doing bulk to depths where it would matter!
